Proposed new error value for implementations that restrict parser ali…

jafingerhut opened the pull request:
…gnment

Similar to the existing statement that some implemtations are allowed
to restrict headers to be multiples of 8 bits long, such
implementations are also likely to impose similar restrictions on the
2-argument extract and advance calls, so that the parsing 'cursor' is
always at a multiple of 8 bits from the beginning of the packet.
These changes propose a way to make that explicit in the
specification.

jafingerhut commented:
Proposed change motivated by implementation work in p4c and BMv2 on advance: https://github.com/p4lang/behavioral-model/issues/727

mbudiu-vmw commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -5021,6 +5021,8 @@ is captured by the following pseudo-code:
 void packet_in.extract<T>(out T headerLvalue,
                           in bit<32> variableFieldSize) {
+   // implementations _may_ include the following line, but need not
+   verify(variableFieldSize[2:0] == 0, error.ParserInvalidAlignment);
```
I think I would prefer this check to be commented-out.

jafingerhut replied:
done

mbudiu-vmw commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -7090,7 +7094,9 @@ error {
     HeaderTooShort,    /// Extracting too many bits into a varbit field.
-    ParserTimeout      /// Parser execution time limit exceeded.
+    ParserTimeout,     /// Parser execution time limit exceeded.
+    ParserInvalidAlignment  /// extract or advance would leave parser with
```
This error name is very specific. How about InvalidArgument, which could probably be reused for other purposes?

jafingerhut replied:
done

jafingerhut commented:
Agreed with all you said in the previous two comments.  No rush on my part to get it merged.  Just wanted to create the PR while the issue was fresh in mind.

jafingerhut commented:
AI Andy: Add this change to the change log at the end of the spec.  Replace "Implementations" with "Targets".

jafingerhut commented:
I believe I was verbally given the go-ahead to merge this in after addressing the comments made during the 2019-Apr-01 LDWG meeting.  If I am mistaken, or further changes or needed, let me know and I can fix things.

jnfoster commented:
👍
